hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -489,13 +489,29 @@ public EntityEntry addEntry(
 
 		final EntityEntry e;
 
-		if( (entity instanceof ManagedEntity) &&  ((ManagedEntity) entity).$$_hibernate_getEntityEntry() != null && status == Status.READ_ONLY) {
-			e = ((ManagedEntity) entity).$$_hibernate_getEntityEntry();
-			e.setStatus( status  );
+		/*
+		Important: 		the following instanceof checks and castings are intentional.
+						Please do *not* refactor to make calls through the EntityEntryFactory
+						interface. Refactoring will results in polymorphic call sites which will
+						severely impact performance.
+		*/
+		if (persister.getEntityEntryFactory() instanceof MutableEntityEntryFactory){
+			e = ((MutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
+					status,
+					loadedState,
+					rowId,
+					id,
+					version,
+					lockMode,
+					existsInDatabase,
+					persister,
+					disableVersionIncrement,
+					lazyPropertiesAreUnfetched,
+					this
+			);
 		}
 		else {
-			final EntityEntryFactory entityEntryFactory = persister.getEntityEntryFactory();
-			e = entityEntryFactory.createEntityEntry(
+			e = ((ImmutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
 					status,
```

Comment:
clearly your results are amazing, so I'll have to believe it :)

Still I'm confused. Let me rephrase that and see if I understand:
with your patch, the type check only happens once at casting and takes 3 bytecodes there; then costs only 1 for each method call. 
W/o your patch, there's no upfront vtable lookup but there's one for each method invocation.
Don't you think that hotspot should be able to rewrite the generated bytecode to keep the first vtable resolution around?

Could you add some notes in the code, or there's risk for these optimisations to be removed in near future.